homeassistant/components/fan/__init__.py

```diff
@@ -347,13 +347,14 @@ def state_attributes(self) -> dict:
             data[ATTR_OSCILLATING] = self.oscillating
 
         if supported_features & FanEntityFeature.SET_SPEED:
-            data[ATTR_PERCENTAGE] = self.percentage
+            if self.percentage is not None:
+                data[ATTR_PERCENTAGE] = self.percentage
```

Comment:
This will cause `percentage` to be missing when its currently unknown. This is not correct, we have to make sure the key is present at any time.

For context see: <https://github.com/home-assistant/architecture/discussions/680#discussion-3724991>

Reply:
Does this mean that it's a valid state for the attributes to be None and thus it's correct to handle on the side where state is reproduced? (The initial direction of this PR) Or is there another layer of indirection somewhere else when translating this to a saved scene?

Reply:
Yeah I'm stuck at this point.

https://github.com/home-assistant/core/pull/62794#issuecomment-1065914641 tells me to not ignore "invalid Scenes" where `percentage: None; preset_mode: str` or `percentage: int; preset_mode: None`.

https://github.com/home-assistant/core/pull/62794#discussion_r846772928 tells me to always have the key present at any time (sometimes as `None`).

I'm more than happy to keep working on the style issues, and figure out how to implement this cleanly, but the fact of the matter is that fan entities with preset modes cannot be added and reproduced onto scenes following the documented design of the fan entity — and I'm afraid I don't have enough time to get involved into redesigning either the Scene handling or the fan entity.

I'd be happy to put a bounty on it if anyone wants to do that! I don't care to be the one fixing fan entities with Scenes, I just want fan entities to work with Scenes!

Reply:
I'm a bit late to the party but I'll try to establish some understanding so we can move this forward:


> [#62794 (comment)](https://github.com/home-assistant/core/pull/62794#issuecomment-1065914641) tells me to not ignore "invalid Scenes" where 

> `percentage: None; preset_mode: str` 

In this state, the fan itself is in control of the percentage since the preset mode ie Auto, etc is controlling it.  The percentage is solely determined by the device and we can never restore it since we don't control it.

> `percentage: int; preset_mode: None`.

In this state the fan is not controlling the percentage and the remote/Home Assistant/physical device/etc is in control.  We can restore the percentage but there is no preset mode set so nothing to restore.

> 
> [#62794 (comment)](https://github.com/home-assistant/core/pull/62794#discussion_r846772928) tells me to always have the key present at any time (sometimes as `None`).


> https://github.com/home-assistant/core/pull/62794#discussion_r846772928 tells me to always have the key present at any time (sometimes as None).

The key should be there if its supported as integrations rely on it.

Reply:
> > `percentage: None; preset_mode: str`
> 
> In this state, the fan itself is in control of the percentage since the preset mode ie Auto, etc is controlling it. The percentage is solely determined by the device and we can never restore it since we don't control it.

Correct.

> 
> > `percentage: int; preset_mode: None`.
> 
> In this state the fan is not controlling the percentage and the remote/Home Assistant/physical device/etc is in control. We can restore the percentage but there is no preset mode set so nothing to restore.

Correct.

> > [#62794 (comment)](https://github.com/home-assistant/core/pull/62794#discussion_r846772928) tells me to always have the key present at any time (sometimes as `None`).
> 
> > [#62794 (comment)](https://github.com/home-assistant/core/pull/62794#discussion_r846772928) tells me to always have the key present at any time (sometimes as None).
> 
> The key should be there if its supported as integrations rely on it.

But that's the thing: those two situations above *cannot ever be restored* — integrations can't rely on them, because the  `fan.set_percentage` and `fan.set_preset_mode` services will reject calls with `None` before they ever reach any integration.

So when restoring a Scene that has `percentage: None; preset_mode: str`, the list of calls currently includes `set_percentage(percentage=None); set_preset(preset_mode='Something')` — the first service call will fail, and the state restore is aborted midway.

That's why my (second) attempt at fixing this with this pull request I changed it to only restore _one_ of the two attributes. I changed due to the comments here to go back to my original suggestion of not storing the None-valued-attribute, since setting `percentage: int` or `preset_mode: str` implies the mutual-exclusive attribute to `None`, but that's what's being rejected here.

Reply:
> But that's the thing: those two situations above _cannot ever be restored_ — integrations can't rely on them, because the `fan.set_percentage` and `fan.set_preset_mode` services will reject calls with `None` before they ever reach any integration.

We can't ever call the services with `None` as these are effectively unknown values.

> 
> So when restoring a Scene that has `percentage: None; preset_mode: str`, the list of calls currently includes `set_percentage(percentage=None); set_preset(preset_mode='Something')` — the first service call will fail, and the state restore is aborted midway.

We shouldn't call it with `None` since the fan is in control and we can't restore what we don't have control of.

> 
> That's why my (second) attempt at fixing this with this pull request I changed it to only restore _one_ of the two attributes. 

That makes sense.  Either we are in control or the fan is in control, its mutually exclusive.

> I changed due to the comments here to go back to my original suggestion of not storing the None-valued-attribute, since setting `percentage: int` or `preset_mode: str` implies the mutual-exclusive attribute to `None`, but that's what's being rejected here.

I think we should still store them all and only restore the one we can control

Reply:
> I think we should still store them all and only restore the one we can control

That works for me (basically, going back to 83892741d6f2eb053d513311c844db4244d61db2.)

If that works for @allenporter and @frenck too, I'm happy to make this change tonight/tomorrow.

Reply:
With all the force pushes, I can't really comment on https://github.com/home-assistant/core/commit/83892741d6f2eb053d513311c844db4244d61db2 

Do you have a branch I could checkout to review?

Reply:
I created a separate pull request (#70294) with a combination of the change from this diff (to use `turn_on` with additional attributes) plus the original change on this request to just ignore either `percentage` or `preset_mode` if they are `None`.

Reply:
I think the version in https://github.com/home-assistant/core/pull/70294 is fine.  It doesn't change the fan entity and handles the mutually exclusive case.

I marked this PR as a draft as i didn't want to close it since it has all the history of how we arrived at #70294